src/dex/idle-dao/utils.ts

```diff
@@ -0,0 +1,215 @@
+  token: string,
+  error_callback?: Function,
+) => {
+  const config = {
+    headers: { Authorization: `Bearer ${token}` },
+  };
+
+  const data = await axios.post(endpoint, {}, config).catch(err => {
+    if (typeof error_callback === 'function') {
+      error_callback(err);
+    }
+  });
+  return data?.data;
+};
+
+export const fetchTokenList_api = async (
+  network: Network,
+  dexHelper: IDexHelper,
+  blockNumber: number,
+  cdoInterface: Interface,
+  erc20Interface: Interface,
+  multiWrapper: MultiWrapper,
+  token: string,
+): Promise<IdleToken[]> => {
+  const data = await getDataWithAuth(endpoints[network], token);
+
+  // Fetch tokenslist from static file
+  if (!data) {
+    return TOKEN_LISTS;
+  }
+
+  const deployedContract = data
+    .filter((d: any) => !!d.cdoAddress)
+    .reduce((acc: any[], d: any) => {
+      if (acc.includes(d.cdoAddress)) return acc;
+      return [...acc, d.cdoAddress];
+    }, []);
+
+  const calls = deployedContract.reduce(
+    (calls: MultiCallParams<any>[], cdoAddress: string) => {
+      calls.push({
+        target: cdoAddress,
+        callData: cdoInterface.encodeFunctionData('AATranche', []),
+        decodeFunction: (
+          result: MultiResult<BytesLike> | BytesLike,
+        ): TrancheToken => {
+          return {
+            cdoAddress,
+            tokenType: 'AA',
+            idleAddress: addressDecode(result),
+          };
+        },
+      });
+
+      calls.push({
+        target: cdoAddress,
+        callData: cdoInterface.encodeFunctionData('BBTranche', []),
+        decodeFunction: (
+          result: MultiResult<BytesLike> | BytesLike,
+        ): TrancheToken => {
+          return {
+            cdoAddress,
+            tokenType: 'BB',
+            idleAddress: addressDecode(result),
+          };
+        },
+      });
+
+      calls.push({
+        target: cdoAddress,
+        callData: cdoInterface.encodeFunctionData('token', []),
+        decodeFunction: (result: MultiResult<BytesLike> | BytesLike): any => {
+          return {
+            cdoAddress,
+            address: addressDecode(result),
+          };
+        },
+      });
+
+      return calls;
+    },
+    [],
+  );
+
+  const results = await multiWrapper.aggregate<Record<string, string>>(
+    calls,
+    blockNumber,
+  );
+
+  const cdosUnderlying: Record<string, string> = results.reduce(
+    (cdosUnderlying, result: any) => {
+      if (!result.address) return cdosUnderlying;
+      return {
+        ...cdosUnderlying,
+        [result.cdoAddress]: result.address,
+      };
+    },
+    {},
+  );
+
+  const idleTokenSymbolsList = await _getIdleTokenSymbols(
+    blockNumber,
+    results
+      .filter(result => !!result.tokenType)
+      .map(result => result.idleAddress),
+    erc20Interface,
+    multiWrapper,
+  );
+
+  const idleTokenDecimals = await _getIdleTokenDecimals(
+    blockNumber,
+    results.filter(result => !result.tokenType).map(result => result.address),
```

Comment:
Incorrect filter? assume it should be `results.filter(result => !!result.tokenType).map(result => result.address)`

Reply:
no, it's correct there, tokens that don't have tokenType would have `address`
tokens with `tokenType`, would have `idleAddress` and `cdoAddress`